### PR TITLE
add act checkpoint at sub layer level

### PR DESCRIPTION
**Pull request by cli99**
This PR allows users to specify what module to do activation checkpoint through fsdp when `activation_checkpointing` is `true` in fsdp_config. It checks the field `activation_checkpointing_target ` in `model_config`, if nothing is specified but `activation_checkpointing` is set to `true`, then `MPTBLOCK` is act checkpointed. To checkpoint the submodules in ATTN_CLASS_REGISTRY, FFN_CLASS_REGISTRY,NORM_CLASS_REGISTRY, add the corresponding names to `activation_checkpointing_target`.
For example, 
```
activation_checkpointing_target: 
    - grouped_query_attention
```
checkpoints the activation of GroupedQueryAttention.
If multiple submodules are up for act checkpointing, add their names to activation_checkpointing_target

### CODE REVIEW

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (mvpatel2000) [inline llmfoundry/models/mpt/modeling_mpt.py:713]: Can we use `list` instead of `lst`

Comment (mvpatel2000) [inline llmfoundry/models/mpt/modeling_mpt.py:713]: why is there a `.replace`?

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (mvpatel2000) [inline llmfoundry/models/mpt/modeling_mpt.py:719]:
```suggestion
                if mod_name.lower() == 'mptblock':
                    mod_type = MPTBlock
```

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (mvpatel2000) [inline llmfoundry/models/mpt/modeling_mpt.py:727]: Can you emit a warning in this case?

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:713]: this list conversion isn't needed. omconfig allows for lists eg adding this in the yaml:
```
activation_checkpointing_target:
  - mptblock
  - grouped_query_attention
```

should result in: 
```config.activation_checkpointing_target = ['mptblock', 'grouped_query_attention']```

Comment (cli99) [reply llmfoundry/models/mpt/modeling_mpt.py:713]: to remove space in the multi-name string, e,g 'grouped_query_attention,  MPTMLP'

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:716]: why cancel out all other typing?

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:728]: this wont work on a list will it?

u might need to do something like
```suggestion
            mod_types = []
            for mod_name in act_ckpt_lst:
                if mod_name.lower() == 'mptblock':
                    mod_types += [MPTBlock]
                elif mod_name in ATTN_CLASS_REGISTRY:
                    mod_types += [ATTN_CLASS_REGISTRY[mod_name]]
                elif mod_name in FFN_CLASS_REGISTRY:
                    mod_types += [FFN_CLASS_REGISTRY[mod_name]]
                elif mod_name in NORM_CLASS_REGISTRY:
                    mod_types += [NORM_CLASS_REGISTRY[mod_name]]
            return isinstance(module, mod_types)
```

this way `activation_checkpointing_target` can actually be a list.

**Review by vchiley — CHANGES_REQUESTED** [review]

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (cli99) [reply llmfoundry/models/mpt/modeling_mpt.py:716]: mptblock is layer-level, other module are sub-layer, if one checkpoints a layer, no need to checkpoint what's inside

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (cli99) [reply llmfoundry/models/mpt/modeling_mpt.py:728]: they are equivalent i think, the return is inside the for loop, once we found a match, then we can return

Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:728]: way I'm seeing it:

if you pass in a list thats say: `[ATTN_CLASS_REGISTRY, NORM_CLASS_REGISTRY]`
in your version, we'll only ever get `return isinstance(module, ATTN_CLASS_REGISTRY)` (even if `isinstance(module, NORM_CLASS_REGISTRY)`

in my version, you'll actually get `isinstance(module, [ATTN_CLASS_REGISTRY,, NORM_CLASS_REGISTRY])`

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:716]: Why not?
It might be a bad idea (we can add a warning if you want), but it should be possible.

eg you can act ckpt the MPTBlock and also the attn_block. why not?

Comment (cli99) [reply llmfoundry/models/mpt/modeling_mpt.py:716]: act ckpt the MPTBlock meaning you recompute the entire block and do not keep the intermediate activations within the block. What does it mean if you act checkpoint attn_block on top of MPTBlock?

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (cli99) [reply llmfoundry/models/mpt/modeling_mpt.py:728]: I see, you are right, thanks!

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (vchiley) [reply llmfoundry/models/mpt/modeling_mpt.py:716]: a block is
-> ln -> Wqkv -> attn -> Wo -> residual_add -> ln -> W1 -> gelu -> W2 -> residual_add ->

act ckpt the block means ONLY the input to that entire block is stored durring fwd pass. Then during the bwd pass, the block does another fwd to recompute intermediate activations

If we act ckpt MPTBlock and also the attn_block, I'd expect the second fwd pass to not store any of the intermediate activations in the attn_block, and then to do a third fwd pass through the attn_block for its bwd. but maybe there this doesn't work for some reason 🤷‍♂️ 

Also image for some reason someone adds more layers (eg attn layers) outside of an MPTBlock and wants to act ckpt those...

we're talking about edge cases i guess. 
either way, either allow for bad behavior (and warn the user that they are doing something weird) or do what you're doing (but again warn the user that you are removing other modules from the act_ckpt list.

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:711]: can we add a `log.info` here saying that it is just act ckpting the mptblock?

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:715]: should we just use the real class names throughout instead of having these lower cased aliases? I think that would be a more clear interface personally.

**Review by dakinggg — COMMENTED** [review]
Two high level comments:
(1) Please add some tests. Not sure how hard it is to test that act ckpt works the way you think it does, but at least test that everything is constructed the way you expect.
(2) Should we add this to all HF models too? I don't see why not. That would go here: https://github.com/mosaicml/llm-foundry/blob/84c86e3b0a3b63c0c71e52f1f762325daa8adc64/llmfoundry/models/hf/hf_fsdp.py#L200

**Comment by cli99** [comment]
> Two high level comments: (1) Please add some tests. Not sure how hard it is to test that act ckpt works the way you think it does, but at least test that everything is constructed the way you expect. (2) Should we add this to all HF models too? I don't see why not. That would go here:
> 
> https://github.com/mosaicml/llm-foundry/blob/84c86e3b0a3b63c0c71e52f1f762325daa8adc64/llmfoundry/models/hf/hf_fsdp.py#L200

(1) are there some existing act ckpt tests to leverage? Or shall we just check to make sure the module names passing works as expected?
(2) sub-block act checkpointing requires we know the module definitions inside the block , it's model dependent.

**Comment by dakinggg** [comment]
(1) Nope. Probably ok to just test that the module passing works as expected. Would be great if we _actually_ tested activation checkpointing, but that does not exist right now, no.
(2) Right, so maybe you have to provide the full import path of the module you want to checkpoint? Is that reasonable?

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:734]: should this be a value or key error?

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (vchiley) [inline llmfoundry/models/mpt/modeling_mpt.py:720]: `log.info` instead of warning?
@dakinggg can verify if thats the correct log level.

Comment (dakinggg) [reply llmfoundry/models/mpt/modeling_mpt.py:720]: I think warning is appropriate, because it is something a user can correct (and probably should)

**Comment by vchiley** [comment]
@dakinggg re (2) this member method is a member method of the MPT model.

enabling this for HF models would requires a custom str to class conversion for a list for every hf model class before it is used in [this call](https://github.com/mosaicml/llm-foundry/blob/84c86e3b0a3b63c0c71e52f1f762325daa8adc64/llmfoundry/models/hf/hf_fsdp.py#L200)...
probably outside the scope of this PR 🤷‍♂️ 
(not sure how `maybe you have to provide the full import path of the module you want to checkpoint` would work)

**Review by vchiley — APPROVED** [review]
lgtm

I'd wait for @dakinggg approval before merging

**Comment by dakinggg** [comment]
@vchiley you would do something like `import('string name of class')` to get the class. but yeah, doesn't need to be part of this PR. @cli99 can you make a jira for adding custom act ckpt to generic HF models and put it in the LLM Foundry backlog jira epic please?

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:714]: Can we make this log a bit more clear by only saying we are ignoring other subblock modules if we actually are (and specifying which ones we're ignoring)?

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (dakinggg) [inline llmfoundry/models/mpt/modeling_mpt.py:731]: can we just print all the options here instead of making the user read code to learn the options?

**Review by dakinggg — COMMENTED** [review]
Also would like basic unit tests testing the use of this param before merging

File: llmfoundry/models/mpt/modeling_mpt.py
Comment (mvpatel2000) [inline llmfoundry/models/mpt/modeling_mpt.py:736]:
```suggestion
                    f'{mod_name} (specified in activation_checkpointing_target) is not a recognized option out of available options {msg}.'
```
nit

**Comment by cli99** [comment]
@dakinggg , added a unit test

File: tests/test_fsdp_act_checkpoint.py
Comment (dakinggg) [inline tests/test_fsdp_act_checkpoint.py:14]: can delete this. the test commands are specified in the makefile

**Review by dakinggg — APPROVED** [review]

**Pull request merged**
